crates/llm-local/src/llama.rs

```diff
@@ -0,0 +1,209 @@
+            let sampling = if temperature <= 0. {
+                Sampling::ArgMax
+            } else {
+                Sampling::TopKThenTopP {
+                    k: params.top_k as usize,
+                    p: params.top_p as f64,
+                    temperature: params.temperature as f64,
+                }
+            };
+            LogitsProcessor::from_sampling(rng.next_u64(), sampling)
+        };
+
+        let mut index_pos = 0;
+        let mut tokens_generated = 0;
+        let mut output_text: String = String::default();
+
+        for index in 0..params.max_tokens {
+            let (context_size, context_index) = if self.cache.use_kv_cache && index > 0 {
+                (1, index_pos)
+            } else {
+                (tokens.len(), 0)
+            };
+            let ctxt = &tokens[tokens.len().saturating_sub(context_size)..];
+            let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
+            let logits = model.forward(&input, context_index, &mut cache)?;
+            let logits = logits.squeeze(0)?;
+            let logits = if params.repeat_penalty == 1. {
+                logits
+            } else {
+                let start_at = tokens
+                    .len()
+                    .saturating_sub(params.repeat_penalty_last_n_token_count as usize);
+                candle_transformers::utils::apply_repeat_penalty(
+                    &logits,
+                    params.repeat_penalty,
+                    &tokens[start_at..],
+                )?
+            };
+            index_pos += ctxt.len();
+
+            let next_token = logits_processor.sample(&logits)?;
+            tokens_generated += 1;
+            tokens.push(next_token);
+
+            // Validate if we have reached the end of the token(s)
+            match eos_token_id {
+                Some(llama::LlamaEosToks::Single(eos_tok_id)) if next_token == eos_tok_id => {
+                    break;
+                }
+                Some(llama::LlamaEosToks::Multiple(ref eos_ids))
+                    if eos_ids.contains(&next_token) =>
+                {
+                    break;
+                }
+                _ => (),
+            }
+            // Decode the token and add it to the output.
+            if let Some(t) = tokenizer.next_token(next_token)? {
+                print!("{}", t);
+                output_text.push_str(&t);
+            }
+        }
+
+        // Decode any remaining tokens after the loop
+        if let Some(rest) = tokenizer.decode_rest()? {
+            output_text.push_str(&rest);
+        }
+
+        Ok(wasi_llm::InferencingResult {
+            text: output_text,
+            usage: InferencingUsage {
+                prompt_token_count: tokens.len() as u32,
+                generated_token_count: tokens_generated,
+            },
+        })
+    }
+}
+
+///  Loads a list of SafeTensors file paths from a given model directory and
+///  path to the model index JSON file.
```

Comment:
The docs don't seem right. The path that gets loaded is the model_dir + the json_file. This makes it sound like `json_file` is an independent path which it is not.

Reply:
updating to mention that it is relative to model directory.